### PR TITLE
Add a pipeline that detects base image updates

### PR DESCRIPTION
**Pull request by damonbarry**
Our core Docker images--Edge Agent, Edge Hub, temp sensor, and the diagnostics module--all target .NET 6.0 Alpine base images. When those images change, often due to important security updates, we need to respond quickly by refreshing our own images. Doing so allows our users to keep their deployments up to date.

We currently have a pipeline that mostly automates this process, but it doesn't automatically detect changes in base images and there are some manually steps like creating release commits and publish GitHub Release pages. As a result, we can't respond as quickly as we'd like when security updates are announced.

This change introduces a pipeline (`detect-image-updates.yaml`) that automatically detects base image updates, and fully automates the release pipeline (`release-core-images.yaml`) for our core images.

The new pipeline is triggered periodically to detect updates in any images we publish from a given branch. The updated "Core Images - Release" pipeline triggers when the detection pipeline completes, and will refresh and release the core images if required.

Prior to this change, the base image update pipeline looked like this:
```
builds/misc/images-release-base-image-update.yaml
  -> builds/e2e/e2e-release.yaml
    -> builds/e2e/templates/e2e-setup-base-image-update-release.yaml
    -> builds/e2e/templates/e2e-run-base-image-update.yaml
builds/misc/images-publish.yaml
```

Note that `images-release-base-image-update.yaml` and the templates listed beneath it are unique to this pipeline; they are not used elsewhere.

The new pipeline looks like this:
```
builds/release/release-core-images.yaml
 -> builds/release/templates/test-core-images.yaml
```

Changes to other files in the repo:
- .`gitignore`: the pipelines live in a new `builds/release` folder, so `.gitignore` was updated to allow git to monitor the folder.
- `builds/e2e/templates/e2e-setup.yaml`: an unused KeyVault secret name was removed.
- `build-rocksdb.yaml`: This was a stage template with a single matrix job. For flexibitilty, it was changed to a job template, and referencing pipelines (`images-release.yaml`, `build-images.yaml`) were updated as well.
- `image-linux.yaml`:
  * Added a better default value for the `bin_dir` parameter, because every use of this template is currently setting the same value.
  * Added the ability to skip downloading consolidated artifacts, in case the artifacts have already been downloaded. This was a cherry-pick from recent changes in main.
  * Added the ability to build the manifest image if a path to the manifest template file is given.
  * Replaced the hard-coded release ACR service connection name with a pipeline variable, which made it easier to test this release pipeline in a non-prod environment.
- `dotnet-code-sign.yaml`: Moved the repetitive and verbose code sign task into its own template.
- `buildBranch.sh`: Added a `--skip-quickstart` option to shorten the build times for scenarios where the long-haul/connectivity test stuff isn't needed. Use of this new flag in the release pipeline is commented out because of a chicken-and-egg problem. We can uncomment the flag when we put out a release that contains these changes.

To test, I ran all the pipelines impacted by this change except `images-release.yaml`. I also ran the new detection and update pipelines many times in a non-prod environment and confirmed the results.

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [X] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [X] Title of the pull request is clear and informative.
- [X] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [X] concise summary of tests added/modified
	- [X] local testing done.

### CODE REVIEW

File: builds/release/detect-image-updates.yaml
Comment (and-rewsmith) [inline builds/release/detect-image-updates.yaml:66]: So we get the base image and the current digest out of some json file? Then get the latest digest from the base image name? Adding more detailed comment as to the approach may help.

File: builds/release/release-core-images.yaml
Comment (and-rewsmith) [inline builds/release/release-core-images.yaml:57]: Why does this imply the release may happen twice? It has something to do with dotnet aspnet vs dotnet runtime, but I don't understand still.

File: builds/release/release-core-images.yaml
Comment (and-rewsmith) [inline builds/release/release-core-images.yaml:117]: What part of this allows devops to be triggered manually? I couldn't point that part out.

File: builds/release/release-core-images.yaml
Comment (and-rewsmith) [inline builds/release/release-core-images.yaml:137]: What happens if one of these isn't done properly? Anything bad?

File: builds/release/templates/test-core-images.yaml
Comment (and-rewsmith) [inline builds/release/templates/test-core-images.yaml:27]: Why does this save us speed and convenience?

File: builds/release/templates/test-core-images.yaml
Comment (and-rewsmith) [inline builds/release/templates/test-core-images.yaml:1]: nit: include this file in PR description

**Review by and-rewsmith — COMMENTED** [review]
It is very cool that we have this!!

Leaving my initial review here with some questions. It would be nice for you to past a link to a passing run. Would help visualize these yamls better for me.

File: builds/release/detect-image-updates.yaml
Comment (damonbarry) [reply builds/release/detect-image-updates.yaml:41]: It's in the [Azure/azure-iotedge](https://github.com/Azure/azure-iotedge/blob/main/product-versions.json) repo.

File: builds/release/detect-image-updates.yaml
Comment (damonbarry) [reply builds/release/detect-image-updates.yaml:66]: The `docker buildx imagetools inspect` command directly examines an image (say, "mcr.microsoft.com/azureiotedge-agent:1.4.2-linux-amd64") and returns JSON (because of the `--format "{{json ...}}"` argument:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "build-arg:EXE_DIR": ".",
    "filename": "Dockerfile",
    "source": "docker/dockerfile:1.4"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "mcr.microsoft.com/dotnet/aspnet:6.0-alpine",
      "pin": "sha256:6032bb4619e4230855986f5ff67ea8eee345a476ab75b46c9977350cd503723a"
    }
  ]
}
```

From `.sources[0].ref` we know which base image Edge Agent targets, and from `.sources[0].pin` we know the digest of the specific image that was used when we built Edge Agent. We use jq to extract those two pieces of information into the variables `$base_image` and `$current_digest`.

We use the `docker buildx imagetools inspect` command again with `$base_image` to get the latest digest for that image. If it doesn't match `$current_digest`, that means the base image has changed since Edge Agent was last released.

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:57]: We use two different .NET base images: temp sensor and diagnostics use dotnet/runtime, while Edge Agent and Edge Hub use dotnet/aspnet. Assuming that those two images (or even platform variations of the same image) aren't published to MCR at _exactly_ the same time, then it's possible that our pipeline will wake up and compare digests while the .NET team is in the middle of updating their images.

As an example, let's say the .NET is in the middle of a release, and their `dotnet/runtime` images have been published to MCR, but their `dotnet/aspnet` images have not (yet). Here's what could happen:
1. Detect-image-updates wakes up, detects that temp sensor and diagnostics images need to be refreshed because their base images have changed
2. Release-core-images is triggered and releases _all_ images, not just temp sensor and diagnostics
4. 12 hours later, Detect-image-updates wakes up and detects that Edge Agent and Edge Hub images need to be refreshed
5. Release-core-images is triggered and releases _all_ images, not just Edge Agent and Edge Hub.

In this case, we'd have two releases (two PATCH version bumps) in less than 24 hours, where one release would have sufficed.

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:117]: This isn't about _allowing_ the pipeline to be triggered manually--in DevOps I think you can trigger any pipeline manually whether it was designed for it or not. This `condition` just passively detects whether the build was triggered manually or not.

I felt it was important to support both types of triggers:
1. [Automatic] Detect-image-updates generates a list of images that need to be updated, and release-core-images wakes up to examine the list, then refreshes/releases the images it is responsible for
2. [Manual] We need to trigger the release manually because:
   1. Detect-image-pipelines is disabled or otherwise not working for some reason
   2. The pipeline ran but encountered an error, so we need to re-run it
   3. Some other reason?

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:137]: Let's say 1.4.2 is the current release, and you want to release 1.4.3.

- If you didn't make a release commit, then the pipeline will detect 1.4.2 as the "new" release and basically re-release it. This technically isn't a great outcome--it means customers could end up with a mix of old and new 1.4.2 images across their devices. They might think all their devices are running the same bits, but technically they're not. The only difference is the base image, so it likely wouldn't be a problem, but theoretically it could lead to confusion if some devices are misbehaving, and others aren't.
- If you made a release commit but didn't tag it, then it's the same as the previous case. We would re-release the 1.4.2.
- If you made a release commit but didn't update the changelog, then we would post the previous release's changelog in the new GitHub release page. Not a big deal. You can always go back and edit the GitHub release page by hand.

File: builds/release/templates/test-core-images.yaml
Comment (damonbarry) [reply builds/release/templates/test-core-images.yaml:27]: Because we can use a 1ES agent (speed because it's faster, convenience because it's stateless) instead of a Raspberry Pi.

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:137]: To mitigate the 1st two bullets from my previous comment, we basically want to add something like the "Are you sure?" dialog box that pops up when you delete a file. 🙂 I see two ways to do this:
1. Add back the "version" pipeline parameter and confirm that it matches the most recent tag in the repo. 
2. Add back an approval step.

Both of these would only happen in the manual case. I think I prefer the 1st option because it's more explicit. If we add an approval step, it would be important for the approver to look at the pipeline output and confirm that the proposed release version matches expectations, but people will probably just press the button without thinking.

Comment (damonbarry) [reply builds/release/release-core-images.yaml:137]: I suppose we could also just refuse to continue a manual pipeline if the most recent tag in the repo matches images that were already published to MCR.

Comment (and-rewsmith) [reply builds/release/release-core-images.yaml:137]: > I suppose we could also just fail the manual pipeline if the most recent tag in the repo matches images that were already published to MCR.

That sounds good to me

**Comment by damonbarry** [comment]
@and-rewsmith almost forgot, links to passing pipeline runs:
[detect-image-updates](https://msazure.visualstudio.com/One/_build/results?buildId=63116099&view=results)
[release-core-images, automatically triggered](https://msazure.visualstudio.com/One/_build/results?buildId=63116228&view=results)
[release-core-images, manually triggered](https://msazure.visualstudio.com/One/_build/results?buildId=63166880&view=results)

File: builds/release/release-core-images.yaml
Comment (and-rewsmith) [inline builds/release/release-core-images.yaml:148]: Won't this already be set from previous step?
https://github.com/Azure/iotedge/pull/6746/files#diff-b64f3d02f0615edcd5ab6f97f28f01823a4131cdb85f8b87e66bc9a27c307900R104

Comment (damonbarry) [reply builds/release/release-core-images.yaml:148]: No, the step you linked is in the `PipelineTrigger` job, which only runs when the pipeline is automatically triggered after detect-image-updates completes. This step is in the `ManualTrigger` job, which is analagous to the `PipelineTrigger` job, but for manually triggered builds.

Comment (damonbarry) [reply builds/release/release-core-images.yaml:148]: I can add a high-level explanation about how the `PrepareRelease` stage (to which both these jobs belong) is stuctured at the top of the stage.

File: scripts/linux/release-tools.sh
Comment (and-rewsmith) [inline scripts/linux/release-tools.sh:107]: clever way to make sure we don't release non-image-update changes 👍

**Review by and-rewsmith — APPROVED** [review]
Looked over the last unseen bit (release-tools.sh). I think it is looking good! I am approving now, but a few things to leave to your judgement:
1. We may want to fail manual pipeline preemptively as discussed.
2. Some of your comments may be best to add as comments in the yamls

File: builds/release/release-core-images.yaml
Comment (and-rewsmith) [reply builds/release/release-core-images.yaml:148]: Ah I was looking at the condition for the step and not the job. I get it.

**Comment by and-rewsmith** [comment]
A couple naming improvements:
1. Filter can be more specific here:
https://msazure.visualstudio.com/One/_build/results?buildId=63166880&view=logs&j=7362d45d-0580-512a-e55b-9fb9f09a18a2&t=7e1c98b8-56ed-5e0b-729e-2ed1908d5bae
2. Job can be more specific here:
https://msazure.visualstudio.com/One/_build/results?buildId=63116099&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:154]: Added a comment to the top `PrepareRelease`.

File: builds/release/release-core-images.yaml
Comment (damonbarry) [reply builds/release/release-core-images.yaml:143]: Added a check to the `ManualTrigger` job that will abort the pipeline if the most recent tag in the repo matches a version that has already been released to MCR.

**Review by and-rewsmith — APPROVED** [review]

**Pull request merged**
